File: src/main/java/de/tectoast/emolga/commands/showdown/SetsCommand.java
package de.tectoast.emolga.commands.showdown;

import de.tectoast.emolga.commands.Command;
import de.tectoast.emolga.commands.CommandCategory;
import de.tectoast.emolga.commands.GuildCommandEvent;
import de.tectoast.emolga.utils.showdown.Analysis;
import de.tectoast.emolga.utils.showdown.Player;
import de.tectoast.emolga.utils.showdown.Pokemon;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;
import java.util.Arrays;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class SetsCommand extends Command {

    private static final Pattern BEGIN_RETURN = Pattern.compile("^\\r\\n");
    private static final Pattern END_RETURN = Pattern.compile("\\r\\n$");
    private final OkHttpClient client = new OkHttpClient().newBuilder().build();

    public SetsCommand() {
        super("sets", "Zeigt die Sets von einem Showdown-Kampf an", CommandCategory.Showdown);
        setArgumentTemplate(ArgumentManagerTemplate.builder()
                .add("url", "Replay-Link", "Der Replay-Link", ArgumentManagerTemplate.Text.any())
                .setExample("!sets https://replay.pokemonshowdown.com/oumonotype-82345404").build());
        beta();
    }

    private static String buildPaste(Player p) {
        return p.getMons().stream().map(SetsCommand::buildPokemon).map(s -> END_RETURN.matcher(BEGIN_RETURN.matcher(s).replaceAll("")).replaceAll("")).collect(Collectors.joining("\r\n\r\n"));
    }

    private static String buildPokemon(Pokemon p) {
        return (p.getNickname().equals(p.getPokemon()) ? p.getPokemon() : p.getNickname() + " (%s)".formatted(p.getPokemon())) + p.buildGenderStr() + (p.getItem().map(s -> " @ " + s).orElse("")) + "  \r\n"
               + "Ability: " + p.getAbility().orElse("unknown") + "  \r\n"
               + p.getMoves().stream().map(s -> "- " + s + "  ").collect(Collectors.joining("\r\n"));
    }

    @Override
    public void process(GuildCommandEvent e) throws Exception {
        String url = e.getArguments().getText("url");
        e.reply(Arrays.stream(new Analysis(url, e.getMessage()).analyse()).map(p -> {
            try {
                String paste = buildPaste(p);
                //e.reply("```" + paste + "```");
                Response res = client.newCall(new Request.Builder()
                        .url("https://pokepast.es/create")
                        .method("POST", new FormBody.Builder()
                                .add("paste", paste)
                                .add("title", "Sets von " + p.getNickname())
                                .add("author", "Emolga")
                                .add("notes", url)
                                .build()
                        )
                        .addHeader("Content-Type", "application/x-www-form-urlencoded")
                        .build()
                ).execute();
                String returl = res.request().url().toString();
                res.close();
                return p.getNickname() + ": " + returl;
            } catch (IOException ex) {
                ex.printStackTrace();
            }
            return "";
        }).collect(Collectors.joining("\n")));
    }
}

File: src/main/kotlin/de/tectoast/emolga/commands/showdown/SetsCommand.kt
@file:Suppress("UNUSED_PARAMETER")

package de.tectoast.emolga.commands.showdown

import de.tectoast.emolga.commands.Command
import de.tectoast.emolga.commands.CommandCategory
import de.tectoast.emolga.commands.GuildCommandEvent
import de.tectoast.emolga.utils.showdown.SDPlayer
import de.tectoast.emolga.utils.showdown.SDPokemon
import java.util.regex.Pattern

@Suppress("unused")
class SetsCommand : Command("sets", "Zeigt die Sets von einem Showdown-Kampf an", CommandCategory.Showdown) {

    init {
        argumentTemplate = ArgumentManagerTemplate.builder()
            .add("url", "Replay-Link", "Der Replay-Link", ArgumentManagerTemplate.Text.any())
            .setExample("!sets https://replay.pokemonshowdown.com/oumonotype-82345404").build()
        beta()
        disable()
    }

    @Throws(Exception::class)
    override suspend fun process(e: GuildCommandEvent) {
        /*val url = e.arguments.getText("url")
        e.reply(Analysis(url, e.message).analyse(e.channel).asFlow().map {
            val paste = buildPaste(it)
            val res = httpClient.post("https://pokepast.es/create") {
                setBody(FormDataContent(Parameters.build {
                    append("paste", paste)
                    append("title", "Sets von ${it.nickname}")
                    append("author", "Emolga")
                    append("notes", url)
                }))
            }
            it.nickname + ": " + res.request.url.toString()
        }.toList().joinToString("\n"))*/
    }

    companion object {
        private val BEGIN_RETURN = Pattern.compile("^\\r\\n")
        private val END_RETURN = Pattern.compile("\\r\\n$")
        private fun buildPaste(p: SDPlayer): String {
            return p.pokemon.map { buildPokemon(it) }.joinToString("\r\n\r\n") {
                END_RETURN.matcher(BEGIN_RETURN.matcher(it).replaceAll("")).replaceAll("")
            }

        }

        private fun buildPokemon(p: SDPokemon): String {
            return buildString {
                /*append(if (p.nickname == p.pokemon) p.pokemon else p.nickname + " (${p.pokemon})")
                append(p.buildGenderStr())
                append(p.item?.let { s: String -> " @ $s" } ?: "")
                append("  \nAbility: ")
                append(p.ability.ifEmpty { "unknown" })
                append("  \n")
                append(p.moves.joinToString("\r\n") { "- $it  " })*/
            }
        }
    }
}
